src/gfn/envs/dag.py

```diff
@@ -0,0 +1,185 @@
+    ):
+        self.num_variables = num_variables
+
+        s0 = torch.zeros(
+            (num_variables, num_variables),
+            dtype=torch.long,
+            device=torch.device(device_str)
+        )
+        sf = torch.ones(
+            (num_variables, num_variables),
+            dtype=torch.long,
+            device=torch.device(device_str)
+        )
+        preprocessor = None  # TODO
+
+        super().__init__(
+            action_space=Discrete(num_variables ** 2 + 1),
+            s0=s0,
+            sf=sf,
+            device_str=device_str,
+            preprocessor=preprocessor
+        )
+
+    def make_States_class(self) -> type[States]:
+        env = self
+
+        class DAGStates(States):
+            state_shape: ClassVar[tuple[int, ...]] = (env.num_variables, env.num_variables)
+            s0 = env.s0
+            sf = env.sf
+
+            @property
+            def forward_masks(self):
+                kwargs = {'dtype': torch.bool, 'device': env.device}
+
+                continue_masks = 1 - (self.states_tensor + self._closure_T)
+                continue_masks = continue_masks.view(*self.batch_shape, -1)
+                continue_masks = continue_masks.to(dtype=torch.bool)
+
+                # The stop action is always a valid action
+                stop_mask = torch.ones(self.batch_shape + (1,), **kwargs)
+                return torch.cat((continue_masks, stop_mask), dim=1)
+
+            @forward_masks.setter
+            def forward_masks(self, value):
+                self._forward_masks = value
+
+            def make_masks(
+                self
+            ) -> Tuple[ForwardMasksTensor, BackwardMasksTensor]:
+                kwargs = {'dtype': torch.bool, 'device': env.device}
+                # Initialize the transitive closure of the transpose of the graphs
+                self._closure_T = torch.zeros(
+                    self.batch_shape + self.state_shape,
+                    **kwargs
+                )
+                self._closure_T.logical_or_(torch.eye(env.num_variables, **kwargs))
+
+                # Initialize the forward masks using `None`. The computation of
+                # the forward masks is being done in the `forward_masks` property
+                forward_masks = None
+
+                # Initialize the backward masks (no valid action)
+                backward_masks = torch.zeros(
+                    self.batch_shape + (env.num_variables ** 2,),
+                    **kwargs
+                )
+                return (forward_masks, backward_masks)
+            
+            def update_masks(self):
+                # TODO: The update of the transitive closure of the graphs
+                # (required to get the forward masks) depends on the last action
+                # applied. Reference: https://github.com/tristandeleu/jax-dag-gflownet/blob/53350bbfba3ab24f9ffa8196a3de42f139d9c5cc/dag_gflownet/env.py#L112-L116
+                self._closure_T = self._closure_T
```

Comment:
@saleml The update of the tensor necessary for computing the forward masks (`_closure_T`) depends on the last action applied (the one that led to the current state). We could implement it by keeping track of the last state inside the state object and reconstructing the action this way, but there may be a cleaner way to do that.

Reply:
Thanks, I'm looking into it now